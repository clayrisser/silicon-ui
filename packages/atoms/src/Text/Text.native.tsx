import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/primitives';
import { TouchableOpacity, ViewProps } from 'react-native';
import { View } from 'dripsy';
import {
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import {
  StyledTextProps,
  StyledViewProps,
  TextProps,
  splitProps
} from './textProps';
import { createStyled } from '../styled';

const StyledView = createStyled<StyledViewProps, ViewProps>(View, {
  dripsy: true
})(compose(color, layout, position, shadow, space, typography));

const StyledText: StyledComponent<
  StyledTextProps,
  StyledTextProps,
  any
> = styled.Text(compose(color, typography, layout));

const Text: FC<TextProps> = (props: TextProps) => {
  const {
    customTextProps,
    nativeTextProps,
    styledTextProps,
    styledViewProps,
    touchableOpacityProps
  } = splitProps(props);
  if (Object.keys(touchableOpacityProps).length) {
    return (
      <TouchableOpacity {...touchableOpacityProps}>
        <StyledView {...styledViewProps} {...nativeTextProps}>
          <StyledText {...styledTextProps}>
            {customTextProps.children}
          </StyledText>
        </StyledView>
      </TouchableOpacity>
    );
  }
  return (
    <StyledView {...styledViewProps} {...nativeTextProps}>
      <StyledText {...styledTextProps}>{customTextProps.children}</StyledText>
    </StyledView>
  );
};

Text.defaultProps = {
  // TODO: lookup fontinfo
  // fontFamily: 'body',
  // fontWeight: 'body',
  // lineHeight: 'body',
  children: '',
  fontSize: 0
};

export default Text;
