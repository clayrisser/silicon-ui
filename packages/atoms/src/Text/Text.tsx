import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import {
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import {
  DetailedHTMLDivProps,
  StyledTextProps,
  TextProps,
  splitProps
} from './textProps';

const HTMLDiv: StyledComponent<
  DetailedHTMLDivProps,
  StyledTextProps,
  object
> = styled.div(compose(color, layout, position, shadow, space, typography));

const Text: FC<TextProps> = (props: TextProps) => {
  const {
    customTextProps,
    styledTextProps,
    styledViewProps,
    touchableOpacityProps
  } = splitProps(props);

  function handleClick(e: any) {
    if (props.onPress) props.onPress(e);
  }

  function handleMouseDown(e: any) {
    if (props.onPressIn) props.onPressIn(e);
  }

  function handleMouseUp(e: any) {
    if (props.onPressOut) props.onPressOut(e);
  }

  return (
    <HTMLDiv
      {...customTextProps}
      {...touchableOpacityProps}
      {...(styledTextProps as any)}
      {...(styledViewProps as any)}
      onClick={handleClick}
      onMouseDown={handleMouseDown}
      onMouseUp={handleMouseUp}
    />
  );
};

Text.defaultProps = {
  children: '',
  fontFamily: 'body',
  fontSize: 0,
  fontWeight: 'body',
  lineHeight: 'body',
  small: 'first'
};

export default Text;
