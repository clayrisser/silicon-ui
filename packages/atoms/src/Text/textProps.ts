import { HTMLAttributes, DetailedHTMLProps } from 'react';
import {
  TouchableOpacityProps,
  TextProps as ReactNativeTextProps
} from 'react-native';
import {
  LayoutProps,
  OpacityProps,
  PositionProps,
  ShadowProps,
  SpaceProps,
  TextColorProps,
  TypographyProps
} from 'styled-system';
import { Theme } from '../themes';

export type DetailedHTMLDivProps = DetailedHTMLProps<
  HTMLAttributes<HTMLDivElement>,
  HTMLDivElement
>;

export interface StyledViewProps
  extends LayoutProps,
    PositionProps,
    OpacityProps,
    ShadowProps,
    SpaceProps {}

export interface StyledTextProps extends TypographyProps, TextColorProps {
  textAlign?: any;
}

export interface NativeTextProps extends ReactNativeTextProps {}

export interface CustomTextProps {
  children?: string;
  theme?: Theme;
  textTransform?: string;
}

export interface TextProps
  extends CustomTextProps,
    NativeTextProps,
    Omit<TouchableOpacityProps, 'style'>,
    StyledTextProps,
    StyledViewProps {}

export const customTextPropsKeys = new Set(['children', 'theme']);

export const touchableOpacityPropsKeys = new Set([
  'activeOpacity',
  'delayLongPress',
  'delayPressIn',
  'delayPressOut',
  'disabled',
  'hitSlop',
  'onBlur',
  'onFocus',
  'onLayout',
  'onLongPress',
  'onPress',
  'onPressIn',
  'onPressOut',
  'style',
  'pressRetentionOffset',
  'testID'
]);

export const styledTextPropsKeys = new Set([
  'color',
  'fontFamily',
  'fontSize',
  'fontStyle',
  'fontWeight',
  'letterSpacing',
  'lineHeight',
  'textAlign'
]);

export const nativeTextPropsKeys = new Set<string>([]);

export interface SplitProps {
  customTextProps: CustomTextProps;
  nativeTextProps: NativeTextProps;
  styledTextProps: StyledTextProps;
  styledViewProps: StyledViewProps;
  touchableOpacityProps: TouchableOpacityProps;
}

export function splitProps(props: TextProps): SplitProps {
  const customTextProps: { [key: string]: any } = {};
  const nativeTextProps: { [key: string]: any } = {};
  const styledTextProps: { [key: string]: any } = {};
  const styledViewProps: { [key: string]: any } = {};
  const touchableOpacityProps: { [key: string]: any } = {};
  Object.entries({ ...props }).forEach(([key, prop]: [string, any]) => {
    if (customTextPropsKeys.has(key)) {
      customTextProps[key] = prop;
    } else if (touchableOpacityPropsKeys.has(key)) {
      touchableOpacityProps[key] = prop;
    } else if (nativeTextPropsKeys.has(key)) {
      nativeTextProps[key] = prop;
    } else if (styledTextPropsKeys.has(key)) {
      styledTextProps[key] = prop;
    } else {
      styledViewProps[key] = prop;
    }
  });
  return {
    customTextProps,
    nativeTextProps,
    styledTextProps,
    styledViewProps,
    touchableOpacityProps
  };
}
