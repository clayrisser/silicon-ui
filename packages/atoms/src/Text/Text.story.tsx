import React from 'react';
import { withKnobs, text, number, color } from '@storybook/addon-knobs';
import Text from './Text';
import Wrapper from '../../storybook/Wrapper';
import storiesOf from '../../storybook/storiesOf';
import withThemeProvider from '../../storybook/withThemeProvider';

storiesOf('Text', module)
  .addDecorator(withKnobs)
  .addDecorator(withThemeProvider)
  .addParameters({
    // docs: { page: docs },
    // jest: ['Button.test.tsx']
  })
  .add('with knobs', () => (
    <Wrapper>
      <Text
        color={color('color', '')}
        fontWeight={text('fontWeight', 'bold')}
        padding={number('padding', 1)}
        margin={number('margin', 1)}
        fontSize={number('fontSize', 2)}
        textAlign={text('textAlign', 'center')}
        opacity={number('opacity', 1)}
        letterSpacing={number('letterSpacing', 0)}
        paddingTop={number('paddingTop', 2)}
        paddingLeft={number('paddingLeft', 1)}
        paddingBottom={number('paddingBottom', 1)}
        paddingRight={number('paddingRight', 2)}
        marginBottom={number('marginBottom', 0)}
        marginTop={number('marginTop', 0)}
        marginLeft={number('marginLeft', 0)}
        marginRight={number('marginRight', 0)}
        // fontFamily={text('font-family', 'Arial, Helvetica, sans-serif')}
      >
        {text('children', 'Hello, world!')}
      </Text>
    </Wrapper>
  ));
