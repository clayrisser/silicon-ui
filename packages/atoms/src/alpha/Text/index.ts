import Text from './Text';
import { TextProps } from './textProps';

export { TextProps };
export default Text;
