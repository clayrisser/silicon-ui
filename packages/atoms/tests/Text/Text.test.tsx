import React from 'react';
// import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Text from '../../src/Text';

describe('<Text />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Text />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('<Text withProps />', () => {
  it('renders with all props', () => {
    const children = 'Text';
    const fontSize = '20px';
    const fontFamily = '';
    const fontWeight = '';
    const lineHeight = '';
    const opacity = '';
    const tree = renderer
      .create(
        <Text
          fontSize={fontSize}
          fontFamily={fontFamily}
          fontWeight={fontWeight}
          lineHeight={lineHeight}
          opacity={opacity}
        >
          {children}
        </Text>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
