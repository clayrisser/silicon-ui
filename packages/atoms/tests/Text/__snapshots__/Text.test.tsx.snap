// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> renders correctly 1`] = `
<div
  className="css-1adg1kh"
  fontFamily="body"
  fontSize={0}
  fontWeight="body"
  onClick={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  opacity="body"
>
  
</div>
`;

exports[`<Text withProps /> renders with all props 1`] = `
<div
  className="css-1joexik"
  fontFamily=""
  fontSize="20px"
  fontWeight=""
  onClick={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  opacity=""
>
  Text
</div>
`;
